src/Workspaces/Core/Desktop/Workspace/Host/TemporaryStorage/TemporaryStorageServiceFactory.cs

```diff
@@ -42,19 +93,69 @@ public ITemporaryTextStorage CreateTemporaryTextStorage(CancellationToken cancel
                 return new TemporaryTextStorage(this);
             }
 
-            public ITemporaryTextStorage AttachTemporaryTextStorage(string storageName, long size, Encoding encoding, CancellationToken cancellationToken)
+            public ITemporaryTextStorage AttachTemporaryTextStorage(string storageName, long offset, long size, Encoding encoding, CancellationToken cancellationToken)
             {
-                return new TemporaryTextStorage(this, storageName, size, encoding);
+                return new TemporaryTextStorage(this, storageName, offset, size, encoding);
             }
 
             public ITemporaryStreamStorage CreateTemporaryStreamStorage(CancellationToken cancellationToken)
             {
                 return new TemporaryStreamStorage(this);
             }
 
-            public ITemporaryStreamStorage AttachTemporaryStreamStorage(string storageName, long size, CancellationToken cancellationToken)
+            public ITemporaryStreamStorage AttachTemporaryStreamStorage(string storageName, long offset, long size, CancellationToken cancellationToken)
+            {
+                return new TemporaryStreamStorage(this, storageName, offset, size);
+            }
+
+            /// <summary>
+            /// Allocate shared storage of a specified size.
+            /// </summary>
+            /// <remarks>
+            /// <para>"Small" requests are fulfilled from oversized memory mapped files which support several individual
+            /// storage units. Larger requests are allocated in their own memory mapped files.</para>
+            /// </remarks>
+            /// <param name="size">The size of the shared storage block to allocate.</param>
+            /// <returns>A <see cref="MemoryMappedInfo"/> describing the allocated block.</returns>
+            private MemoryMappedInfo CreateTemporaryStorage(long size)
+            {
+                if (size >= SingleFileThreshold)
+                {
+                    // Larger blocks are allocated separately
+                    var mapName = CreateUniqueName(size);
+                    var storage = MemoryMappedFile.CreateNew(mapName, size);
+                    return new MemoryMappedInfo(new ReferenceCountedDisposable<MemoryMappedFile>(storage), mapName, 0, size);
+                }
+
+                lock (_gate)
+                {
+                    // Obtain a reference to the memory mapped file, creating one if necessary. If a reference counted
+                    // handle to a memory mapped file is obtained in this section, it must either be disposed before
+                    // returning or returned to the caller who will own it through the MemoryMappedInfo.
+                    var reference = _fileReference.TryAddReference();
+                    if (reference == null || _offset + size > _fileSize)
```

Comment:
this seems weird.  say the reference is non-null.  You've now incremented the refcount.  But hten you check the size stuff.  And if the second condition is true, you decide to overwrite reference and _fileReference.  But nothing seems to decrement the ref count for _fileReference that you just incremented here.